### DISCUSSION TITLE
What's the difference between would_apply and apply?

### DISCUSSION THREAD

**Discussion by sirdody** [harfbuzz/harfbuzz#3775] [General] [answered]
Hi behdad,

Can you help me differentiate between would_apply and apply (would_match_input, match_input, and other similar ones) functions? Why do we need would_apply function? I saw it's only called in Indic shaper. 

The others are match_input and match_lookahead. The match_input is similar to match_lookahead with some differences (compute total_component_count, match position). Can you clarify me the intention of each of them, and the difference between them?

Thank you a lot

**Comment by behdad** [comment, marked as answer]
As the names suggest, one just checks whether a lookup would apply to a glyph sequence, while the other actually applies it. The `would_apply` is needed in the Indic shaper because the Indic shaper rules require such introspection into the font lookups for reclassification of the consonants into below- or post-forms.
